monkey/monkey_island/cc/services/agent_plugin_service/mongo_agent_plugin_repository.py

```diff
@@ -0,0 +1,226 @@
+
+        for manifest_dict in manifest_dicts:
+            try:
+                manifest = AgentPluginManifest(**manifest_dict["plugin_manifest"])
+            except Exception as err:
+                raise RetrievalError(
+                    f"Can't create plugin manifest from the data in the database: {err}"
+                ) from err
+            plugin_type = manifest.plugin_type
+            plugin_name = manifest.name
+            if plugin_type not in manifests:
+                manifests[plugin_type] = {}
+            manifests[plugin_type][plugin_name] = manifest
+
+        return manifests
+
+    def store_agent_plugin(self, operating_system: OperatingSystem, agent_plugin: AgentPlugin):
+        plugin_name = agent_plugin.plugin_manifest.name
+        plugin_type = agent_plugin.plugin_manifest.plugin_type
+        try:
+            plugin_dict = self._get_agent_plugin(plugin_type, plugin_name)
+        except Exception:
+            plugin_dict = self._encode_agent_plugin(agent_plugin)
+            plugin_dict[BINARY_OS_MAPPING_KEY] = {}
+
+        if operating_system.value in plugin_dict[BINARY_OS_MAPPING_KEY]:
+            raise StorageError(
+                f"Plugin {plugin_name} of type {plugin_type} already has a binary for "
+                f"operating system {operating_system}"
+            )
+
+        try:
+            id = self._agent_plugins_binaries_collections[operating_system].put(
+                agent_plugin.source_archive
+            )
+            plugin_dict[BINARY_OS_MAPPING_KEY][operating_system.value] = id
+        except Exception:
+            raise StorageError(
+                f"Failed to store binary for plugin {plugin_name} of type {plugin_type} for "
+                f"operating system {operating_system}"
+            )
+
+        print(f"Writing plugin {plugin_dict}")
+        self._agent_plugins_collection.update_one(
+            {"plugin_manifest.name": plugin_name, "plugin_manifest.plugin_type": plugin_type.value},
+            {"$set": plugin_dict},
+            upsert=True,
+        )
+
+    def _encode_agent_plugin(self, agent_plugin: AgentPlugin) -> Dict[str, Any]:
+        dict = agent_plugin.dict(simplify=True, exclude={"source_archive"})
+
+        return self._encode_dict(dict)
+
+    def _encode_dict(self, d: dict):
+        new_dict = {}
+        for k, v in d.items():
+            if isinstance(k, Enum):
+                k = k.value
+            if isinstance(v, Enum):
+                v = v.value
+            if isinstance(v, dict):
+                v = self._encode_dict(v)
+
+            new_dict[k] = v
+
+        return new_dict
+
+    def remove_agent_plugin(
+        self,
+        agent_plugin_type: AgentPluginType,
+        agent_plugin_name: str,
+        operating_system: Optional[OperatingSystem] = None,
+    ):
+        try:
+            plugin_dict = self._get_agent_plugin(agent_plugin_type, agent_plugin_name)
+        except Exception:
```

Comment:
Is this too broad? If there's a connection error, this should be wrapped as a `RemovalError`. If there's a "not found" error from Mongo, then this logic is correct.

Reply:
Is it necessary, though? Why don't we just call `_remove_agent_plugin()` instead of checking whether or not the plugin exists?

Reply:
This is still likely too broad. It needs to distinguish between `UnknownRecordError` (NOP) and any other error.